qiskit/transpiler/target.py

```diff
@@ -276,6 +278,37 @@ def __init__(
         self._coupling_graph = None
         self._instruction_durations = None
         self._instruction_schedule_map = None
+        self._non_global_basis = None
+        self._non_global_basis_strict = None
+
+    def get_non_global_operation_names(self, strict_direction=False):
+        """Return the non-global operation names for the target
+
+        The non-global operations are those in the target which don't apply
+        on all qubits (for single qubit operations) or all multi-qubit qargs
+        (for multi-qubit operations).
+
+        Args:
+            strict_direction (bool): If set to ``True`` the multi-qubit
+                operations considered as non-global respect the strict
+                direction (or order of qubits in the qargs is significant). For
+                example, if ``cx`` is defined on ``(0, 1)`` and ``ecr`` is
+                defined over ``(1, 0)`` by default neither would be considered
+                non-global, but if ``strict_direction`` is set ``True`` both
+                ``cx`` and ``ecr`` would be returned.
+
+        Returns:
+            List[str]: A list of operation names for operations that aren't global in this target
+        """
+        if strict_direction:
+            if getattr(self, "_non_global_basis_strict", None) is None:
+                self._non_global_basis_strict = super()._get_non_global_operation_names(
+                    strict_direction
+                )
+            return self._non_global_basis_strict
+        if getattr(self, "_non_global_basis", None) is None:
```

Comment:
Is there a reason to use `getattr` over just this? 
```suggestion
            if self._non_global_basis_strict is None:
                self._non_global_basis_strict = super()._get_non_global_operation_names(
                    strict_direction
                )
            return self._non_global_basis_strict
        if self._non_global_basis is None:
```

Reply:
There were code paths that were apparently skipping `__init__` that called this method (given how subclassing works with PyO3 I could see that being the case). When I wrote this locally I had it as you suggested and was getting `AttributeErrors` in a few tests because the attribute hadn't been populated yet. Doing the getattr covered that edge case.

Reply:
It might be nice to leave a comment explaining this for our future selves 😄

Reply:
or possibly fix the places that are calling methods that require initialisation without actually initialising the objects... That's a huge footgun waiting to go off further.

Reply:
@mtreinish which tests are failing? Is it possible that it might have something to do with `FakeTarget`? I don't think so since it calls `__init__`, but it doesn't call `__new__`.

Reply:
It were some transpile tests that were failing in the basis translator. I think it was related to the `_FakeTarget`. I fixed the issue in: https://github.com/Qiskit/qiskit/pull/14921/commits/19b890bf67ef7f4979958028537c0175742c69ec and used the more normal syntax without the getattr in that commit.